qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,625 @@
+                    function {}() {{
+                        var res = {};
+                        if (res instanceof Error) {{
+                            throw res;
+                        }} else {{
+                            return res;
+                        }}
+                    }}
+                """
+                call_str = call_str_f.format(name)
+                eval_str = decl_str_f.format(name, call_str)
+                self._engine.evaluate(eval_str)
+
+        self._engine.evaluate(pac_str, "pac")
+        global_js_object = self._engine.globalObject()
+        self._resolver = global_js_object.property("FindProxyForURL")
+        if not self._resolver.isCallable():
+            err = "Cannot resolve FindProxyForURL function, got '{}' instead"
+            raise EvalProxyError(err.format(self._resolver.toString()))
+
+    def resolve(self, query):
+        """Resolve a proxy via PAC.
+
+        Args:
+            query: QNetworkProxyQuery.
+
+        Return:
+            A list of QNetworkProxy objects in order of preference.
+        """
+        result = self._resolver.call([query.url().toString(),
+                                      query.peerHostName()])
+        result_str = result.toString()
+        if not result.isString():
+            err = "Got strange value from FindProxyForURL: '{}'"
+            raise EvalProxyError(err.format(result_str))
+        return self._parse_proxy_string(result_str)
+
+
+class PACFetcher(QObject):
+
+    """Asynchronous fetcher of PAC files."""
+
+    def __init__(self, url, parent=None):
+        """Resolve a PAC proxy from URL.
+
+        Args:
+            url: QUrl of a PAC proxy.
+        """
+        pac_prefix = "pac+"
+
+        super().__init__(parent)
+
+        assert url.scheme().startswith(pac_prefix)
+        url.setScheme(url.scheme()[len(pac_prefix):])
+
+        self._manager = QNetworkAccessManager()
+        self._manager.setProxy(QNetworkProxy(QNetworkProxy.NoProxy))
+        self._reply = self._manager.get(QNetworkRequest(url))
+        self._reply.finished.connect(self._finish)
+        self._pac = None
+        self._loop = None
+
+    @pyqtSlot()
+    def _finish(self):
+        if self._reply.error() != QNetworkReply.NoError:
+            log.network.error("Can't fetch PAC file from URL")
+        else:
+            try:
+                pacscript = bytes(self._reply.readAll()).decode()
+            except UnicodeError:
+                log.network.exception("Invalid encoding of a PAC file")
+            try:
+                self._pac = PACResolver(pacscript)
+                log.network.debug("Successfully evaluated PAC file.")
+            except EvalProxyError as e:
+                error = "Error in PAC evaluation: {!s}.".format(e)
+                log.network.exception(error)
+        if self._loop is not None:
+            self._reply.deleteLater()
+        else:
+            self._manager = None
+            del self._reply
```

Comment:
`del ...` doesn't give you any guarantee it's actually deleted from Qt's point of view though, all it does is removing that one Python reference, but you don't know when the garbage collector will kick in (or if there are any other references).

Can't we find some better way to exit the event loop?